### DISCUSSION TITLE
#15 - CI/💿 Workflows

### DISCUSSION THREAD

**Discussion by Annalytic-programming** [CUMGroup/AMOGUS#125] [Blog]
# Heio heio heio 👋,

![](https://media.tenor.com/EfWPMN961jUAAAAC/funny-chicken.gif)

This week we were talking about CI/CD in our Softwareengineering lecture.
As we already finished our CI/CD workflow before this lecture, we want to present you our CI/CD solution.

## Continious Integration Workflow

Whenever a pull request to the main branch changes the `./src/` folder 📁 our CI Workflow gets triggered. This includes:

### Building the API and UI
[This configuration file](https://github.com/CUMGroup/AMOGUS/blob/main/.github/workflows/build-on-pullrequest.yaml) triggers Github Actions to checkout our project and build the .NET project as well as the Angular project. If only one fails the workflow fails and the pull request gets 
![](https://media.tenor.com/YFhAP8U26GQAAAAM/rejected-stamp.gif)

### Run all Unit Tests
Look [here](https://github.com/CUMGroup/AMOGUS/blob/main/.github/workflows/test-on-pullrequest.yaml) to see the config file for running all our [unit tests](https://cumgroup.github.io/AMOGUS/Developer%20Docs/4.%20Testing/Testplan/). If any test fails the pull request gets rejected and the workflow fails. ❌

The cherry on top 🍒 is that a [code coverage report](https://cumgroup.github.io/AMOGUS/assets/images/cicd/coverage.png) is also generated.

## Continous Deployment Workflow

If our pull request gets accepted, we can go on pushing the changes to the main branch. Every time something is pushed to the main branch our CD Workflow gets triggered. This includes two workflows:

### API Deployment
[This](https://github.com/CUMGroup/AMOGUS/blob/main/.github/workflows/deploy-api.yaml) builds a docker image which is then pushed to our DockerHub 🐋 Repository and sends a webhook 🎣 to the deployment script on our server.

### UI Deployment
The UI workflow works the same. [This](https://github.com/CUMGroup/AMOGUS/blob/main/.github/workflows/deploy-ui.yaml) is the corresponding config file.

## Serverside

Last but not least every time a request comes in, the webhook 🎣 on the server starts the build script which pulls the docker image form DockerHub, 🐳 stops the currently running container then prunes all remaining artifacts 📄 and runs the new container with all required environment variables.

If you want futher information, you can also read our corresponding CI/CD Documentation [here](https://cumgroup.github.io/AMOGUS/Developer%20Docs/6.%20CICD/CICDDoc/)

# Benefits from our workflow

Because of just using Github Actions and basic Docker configurations we have a minimal setup. We don't need to setup Kubernetes or other container management systems.

Btw. we hope you get your [AMOGUS](https://amogus.alexmiha.de/) streaks up :D

Yours truly,
CUMGroup

**Comment by vans1646** [comment]
Hi CUMGroup,
great to read another new blog post from you!

I didn't know that you had an accessable Documentation, so I had a look at that and it's pretty cool, nicely done. Therefore everything was very understandable and I think you're taking the right path with your approach.

I also had a look at your .yaml files, but since we also use GitHub Actions these look pretty similar so no questions arose from my side.

The only thing I'm wondering about: do you also do an internal review for every pull request or do you compeletely rely on your test results?

Keep the good work up!

Cheers,
Capyclue

**Reply by AlexMi-Ha** [reply]
Hi Capyclue,
thanks for the regular feedback! ^^

Yup our [documentation](http://amogus-docs.alexmiha.de) is neatly contained withing a website generated by mkDocs, where you can generate a website from Markdown files in your repo.

Yes, we are constantly reviewing each others pull requests! Per main-branch protection rules at least on reviewer is required. That way we can guarantee better code and less bugs.

Yours truly,
CUMGroup

**Comment by schw1erig** [comment]
Hi AMOGUS,
You've made nice progress with implementing a continuous integration workflow. The choice of Github Actions for CI provides a high-powered automation tool for automatic testing.
That the automatic generated report gives instant feedback is a good way to avoid waiting for a manual review and insures a faster workflow.
I'm looking forward to your future updates!
Best regards
The Common Team

**Reply by AlexMi-Ha** [reply]
Hi Common,
yes our automated workflows give fast feedback. Thats really neat.
BUT of course we still have to wait for at least one reviewer to look at pull requests before we can merge it to main.
That way we have less of a risk of little bugs and code smells!

Yours truly,
CUMGroup

**Comment by Richone01** [comment]
Hi CUMGroup,

I really like your CI/CD workflow, especially that it's so minimal. All through the article I've been like "what tools do they use?". But it's all just Github actions, didn't know they're that powerful.  Are your unit test results also documented somewhere or does it work like "tests passed, the rest doesn't matter"?

Keep up the good work!
Leon from DHBWorkout

**Reply by Richone01** [reply]
My bad, just seen that there's a cherry on top xD

**Reply by AlexMi-Ha** [reply]
Hi DHBWorkout,
our goal was to keep it as minimal as possible to reduce the setup work but still ensure a solid cicd process!

As you've already found, our test results are documented in our github actions logs, but for future projects I really want to make it easier to access coverages and more importantly coverage deltas. I've already looked at [CodeCov](https://about.codecov.io/) as suggested by Team FiBo, but for this project it's out of scope ^^

Yours truly,
CUMGroup

**Comment by mikaWohlfart** [comment]
Hi CUMGroup,

I'm excited to hear about your CI/CD workflow. I'm impressed with how you've used GitHub Actions and Docker to create a solution that is both simple and effective.

I like how you've used GitHub Actions to automate the build and test process. This frees up your team to focus on other tasks, and it helps to ensure that your code is always up-to-date and free of errors.

Sincerely,
Framgang
